src/login.rs

```diff
@@ -182,20 +211,69 @@ impl Login {
 
         let self_ = imp::Login::from_instance(self);
         let visible_page = self_.content.visible_child_name().unwrap();
-        if visible_page == "phone-number-page" {
-            let encryption_key = self_.custom_encryption_key_entry.text().to_string();
-            if !encryption_key.is_empty() {
-                self.change_encryption_key();
+
+        match visible_page.as_str() {
+            "phone-number-page" => {
+                let encryption_key = self_.custom_encryption_key_entry.text().to_string();
+                if !encryption_key.is_empty() {
+                    self.change_encryption_key();
+                }
+
+                self.send_phone_number();
+            }
+            "code-page" => self.send_code(),
+            "registration-page" => {
+                if !self_.show_tos_popup.get() || self_.tos_accepted_by_user.get() {
+                    // Just proceed if the user either doesn't need to accept the ToS or has
+                    // already accepted it
+                    self.send_registration()
+                } else {
+                    // Force the ToS dialog for the user before he can proceed
+                    self.show_tos_dialog(true);
+                }
             }
+            "password-page" => self.send_password(),
+            "encryption-key-page" => self.send_encryption_key(false),
+            other => unreachable!("no page named '{}'", other),
+        }
+    }
+
+    fn show_tos_dialog(&self, user_wants_to_proceed: bool) {
+        let self_ = imp::Login::from_instance(self);
 
-            self.send_phone_number();
-        } else if visible_page == "code-page" {
-            self.send_code();
-        } else if visible_page == "password-page" {
-            self.send_password();
-        } else if visible_page == "encryption-key-page" {
-            self.send_encryption_key(false);
+        let builder = gtk::MessageDialog::builder()
+            .use_markup(true)
+            .secondary_text(&*self_.tos_text.borrow())
+            .modal(true)
+            .transient_for(self.root().unwrap().downcast_ref::<gtk::Window>().unwrap());
+
+        let dialog = if self_.show_tos_popup.get() {
+            builder
+                .buttons(gtk::ButtonsType::YesNo)
+                .text(&gettext("Do You Accept the Terms of Service?"))
+        } else {
+            builder
+                .buttons(gtk::ButtonsType::Ok)
+                .text(&gettext("Terms of Service"))
         }
+        .build();
+
+        dialog.run_async(clone!(@weak self as obj => move |dialog, response| {
+            let self_ = imp::Login::from_instance(&obj);
+
+            if matches!(response, gtk::ResponseType::No) {
+                self_.tos_accepted_by_user.set(false);
+                if user_wants_to_proceed {
```

Comment:
This should not be needed anymore if you apply the change of the review above.

Reply:
And ofc github mixed up the reviews :D

I was referring to the review of line 250.

Reply:
At least the `obj.unfreeze();` is needed here, otherwise the UI becomes frozen after the user decline the dialog.

Reply:
@melix99 sorry, github loaded your second comment after I commented. I am a bit confused now. :smiley: 
Otherwise, everything is fine with this if clause?

Reply:
I was referring to the fact that the if was not needed because both the Yes and No response can be received only when `user_needs_to_accept` is true, but I noticed that you left the check in the Yes response, maybe because for some reason the Yes response gets activated even when pressing the OK button?

Reply:
Oh, I see that I have overseen a ` && user_needs_to_accept` that I have removed now. I Also left some comments. Feel free to reopen if something else is missing.